proxy/coco/src/peer/subroutines.rs

```diff
@@ -0,0 +1,280 @@
+    subscriber: broadcast::Sender<Event>,
+}
+
+impl Subroutines {
+    /// Constructor.
+    pub fn new(
+        state: State,
+        store: kv::Store,
+        run_config: RunConfig,
+        protocol_events: BoxStream<'static, ProtocolEvent<Gossip>>,
+        subscriber: broadcast::Sender<Event>,
+        control_receiver: mpsc::Receiver<control::Request>,
+    ) -> Self {
+        let announce_timer = interval(run_config.announce.interval);
+        let waiting_room_timer = interval(run_config.waiting_room.interval);
+
+        let (input_sender, inputs) = mpsc::channel::<Input>(RECEIVER_CAPACITY);
+
+        let run_state = RunState::from(run_config);
+
+        let inputs = {
+            let mut coalesced = SelectAll::new();
+            coalesced.push(protocol_events.map(Input::Protocol).boxed());
+            coalesced.push(
+                announce_timer
+                    .map(|_tick| Input::Announce(AnnounceInput::Tick))
+                    .boxed(),
+            );
+            coalesced.push(
+                waiting_room_timer
+                    .map(|_tick| Input::Request(RequestInput::Tick))
+                    .boxed(),
+            );
+            coalesced.push(
+                control_receiver
+                    .map(|request| match request {
+                        control::Request::Urn(urn, time, sender) => {
+                            Input::Request(RequestInput::Requested(urn, time, Some(sender)))
+                        },
+                    })
+                    .boxed(),
+            );
+            coalesced.push(inputs.boxed());
+
+            coalesced
+        };
+
+        Self {
+            pending_tasks: FuturesUnordered::new(),
+            inputs,
+
+            state,
+            store,
+            run_state,
+
+            subscriber,
+            input_sender,
+        }
+    }
+}
+
+impl Drop for Subroutines {
+    fn drop(&mut self) {
+        for task in self.pending_tasks.iter_mut() {
+            task.abort()
+        }
+    }
+}
+
+impl Future for Subroutines {
+    type Output = Result<(), spawn_abortable::Error>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        // Drain the task queue
+        loop {
+            match self.pending_tasks.poll_next_unpin(cx) {
+                Poll::Ready(Some(Err(e))) => {
+                    log::warn!("error in spawned subroutine task: {:?}", e);
+                    return Poll::Ready(Err(e));
+                },
+                Poll::Ready(Some(Ok(()))) => continue,
+                // Either pending, or FuturesUnordered thinks it's done, but
+                // we'll enqueue new tasks below
+                Poll::Ready(None) | Poll::Pending => break,
+            }
+        }
+
+        loop {
+            match self.inputs.poll_next_unpin(cx) {
+                Poll::Ready(Some(input)) => {
+                    log::debug!("handling subroutine input: {:?}", input);
```

Comment:
For understanding this method and keeping it focused on the concurrency logic it might be beneficial to extract this block into a separate function.

Reply:
To clarify are you suggesting to introduce something like `handle_input` on `Subroutines` to reduce the size of this method body?